File: java/amp-rabbit/src/main/java/amp/rabbit/topology/RoutingInfo.java
package amp.rabbit.topology;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * RoutingInfo is the information needed by the Topology Client to creating connections
 * to brokers, and bind to exchanges and queues.
 * 
 * RoutingInfo includes both the information to produce, as well as, consume messages.  We
 * have made the decision to separate these concepts because in our model, we believe that
 * you may produce on one route and consume from another for security reasons.
 * 
 * We also force client to accept a collection of producing and consuming routes.  The intent
 * is that you may need to push/pull messages to/from multiple clusters based on topology.
 * 
 * @author Richard Clayton (Berico Technologies)
 */
public class RoutingInfo {

    protected ArrayList<ProducingRoute> producingRoutes = 
    		new ArrayList<ProducingRoute>();
    protected ArrayList<ConsumingRoute> consumingRoutes = 
    		new ArrayList<ConsumingRoute>();
    
    public RoutingInfo(){}
    public RoutingInfo(Collection<ProducingRoute> producingRoutes, 
    				   Collection<ConsumingRoute> consumingRoutes){
    	
    		this.producingRoutes.addAll(producingRoutes);
    		this.consumingRoutes.addAll(consumingRoutes);
    }
    
	public List<ProducingRoute> getProducingRoutes() {
		return producingRoutes;
	}
	
	public List<ConsumingRoute> getConsumingRoutes() {
		return consumingRoutes;
	}
	
	public List<BaseRoute> getAllRoutes() {
		ArrayList<BaseRoute> routes = new ArrayList<BaseRoute>();
		routes.addAll(producingRoutes);
		routes.addAll(consumingRoutes);
		return routes;
	}
	
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result
				+ ((consumingRoutes == null) ? 0 : consumingRoutes.hashCode());
		result = prime * result
				+ ((producingRoutes == null) ? 0 : producingRoutes.hashCode());
		return result;
	}
	
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		RoutingInfo other = (RoutingInfo) obj;
		if (consumingRoutes == null) {
			if (other.consumingRoutes != null)
				return false;
		} else if (!consumingRoutes.equals(other.consumingRoutes))
			return false;
		if (producingRoutes == null) {
			if (other.producingRoutes != null)
				return false;
		} else if (!producingRoutes.equals(other.producingRoutes))
			return false;
		return true;
	}
}

File: dotnet/source/amp.rabbit/topology/RoutingInfo.cs
using System.Collections.Generic;
using amp.utility;

namespace amp.rabbit.topology
{
    public class RoutingInfo
    {
        public IEnumerable<ProducingRoute> ProducingRoutes { get; private set; }
        public IEnumerable<ConsumingRoute> ConsumingRoutes { get; private set; }

        public RoutingInfo(IEnumerable<ProducingRoute> producingRoutes, IEnumerable<ConsumingRoute> consumingRoutes)
        {
            ProducingRoutes = producingRoutes;
            ConsumingRoutes = consumingRoutes;
        }

        protected bool Equals(RoutingInfo other)
        {
            return HashUtility.Compare(ProducingRoutes, other.ProducingRoutes) 
                && HashUtility.Compare(ConsumingRoutes, other.ConsumingRoutes);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((RoutingInfo) obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return ((ProducingRoutes != null ? HashUtility.GetHashCode(ProducingRoutes) : 0)*397) 
                    ^ (ConsumingRoutes != null ? HashUtility.GetHashCode(ConsumingRoutes) : 0);
            }
        }
    }
}
